core/shared/src/main/scala/zio/Schedule.scala

```diff
@@ -174,7 +196,7 @@ trait Schedule[-Env, -In, +Out] extends Serializable { self =>
   def ||[Env1 <: Env, In1 <: In, Out2](that: Schedule[Env1, In1, Out2])(implicit
     zippable: Zippable[Out, Out2]
   ): Schedule.WithState[(self.State, that.State), Env1, In1, zippable.Out] =
-    (self unionWith that)((l, r) => Schedule.minOffsetDateTime(l, r))
+    (self unionWith that)((l, r) => (l union r).getOrElse(l min r))
```

Comment:
I think we probably need a richer representation of `Interval` that allows us to represent disjoint unions. For example, if we have `Schedule.dayOfWeek(2) || Schedule.dayOfWeek(4)` we want to preserve the full information that the schedule is willing to recur on either day because we may later intersect it within a schedule that is only willing to recur on the later day.

Reply:
Another approach is to keep track of something like "current interval", which would be the interval that either overlaps with the current moment in time (if there is such an interval), or the first interval after the current moment in time (if there is no such interval). If you are keeping track of the "current interval" on both sides for an intersection or union, then you can make a decision about the union or intersection without first performing an interval union / intersection. Basically, the current intervals become the "system of record" and decisions about the next schedule interval occur based on those.